pyk/src/pyk/kast/att.py

```diff
@@ -123,6 +123,11 @@ def pretty(self, value: str) -> str:
 
 class LocationType(AttType[tuple[int, int, int, int]]):
     def from_dict(self, obj: Any) -> tuple[int, int, int, int]:
+        if isinstance(obj, str):
```

Comment:
There should be a single JSON representation for each attribute. Which component emits this format?

Reply:
The `Att` class in `outer_syntax` currently only accepts `tuple[str,str]` for its items, so I'm encoding the location in a string. The items get turned into a dictionary and passed here when converting to a `KAtt` so I needed logic to decode it.

Reply:
You can handle this by adding a new method to `AttType`:

```python3
@abstractmethod
def parse(self, text: str) -> T: ...
```

(And consider renaming `pretty` to `unparse` to maintain symmetry.)

Reply:
Do you think the check and call to `parse` should happen in the `AttType`'s `from_dict` method, or in `KAtt`'s `from_dict`?

Reply:
You can call this when you convert `outer_syntax.Att` to `KAtt`: you look up `AttKey` based on the left element of the tuple, then parse the right element using the corresponding `AttType`.

Reply:
There's a problem here. `LocationType.parse` needs to return `tuple[int, int, int, int]` (because that's what the type variable `T` is assigned to in the superclass), so its result can't be passed back in to `LocationType.from_dict`, it will expect a `list`.

Reply:
There's no need to call `from_dict`, you can construct the `AttEntry` directly.

Reply:
Oh, I think I see. `parse` can return the entire `AttType[T]` and not just a `T`

Reply:
nvm. I just confused myself there.

But what about what I have implemented right now, where `KAtt.from_dict` picks between calling `from_dict` and `parse` depending on the type of the value? It seems like a pretty sensible place to determine when to parse a string and it's localized to `att.py`

Reply:
`from_dict` is for parsing the JSON format, i.e. `compiled.json`. Parsing attributes from the textual format is a new use case. In my opinion, adding a new method is the right thing to do.